### ISSUE TITLE
The performance has deteriorated significantly after fine-tuning with TRL but increase while using llama-factory

### ISSUE THREAD

**Issue by Hasuer** [huggingface/trl#3432]
I used exactly the same data, base model, and identical configurations for SFT and evaluated by `lighteval`. However, models fine-tuned using different frameworks showed significant performance differences. Specifically, the model fine-tuned with TRL performs significantly worse, while the one fine-tuned with LLaMA-Factory performs much better.

Below are the evaluation results:

AIME24 (n=32) | AIME25 (n=32) | Math-500 | GPQA
-- | -- | -- | --
Backbone | 72.6 | NaN | 94.3
TRL | 13.96 | 11.35 | 77.2
LLaMA-Factory | 75.94 | 63.54 | 97.6

The dataset used is [s1k_1.1](https://huggingface.co/datasets/simplescaling/s1K-1.1/viewer/default/train), which is the version with DeepSeek-R1 answers.

The fine-tuned model is `DeepSeek-R1-Distill-Qwen-32B`.

For TRL, I used the model's default chat-template, which removes thinking trajectory, and the data was converted into a [prompt-completion format](https://github.com/huggingface/trl/blob/main/docs/source/dataset_formats.md) and train with `completion_only_loss=True`.

For LLaMA-Factory, I used the latest version and applied the `deepseekr1` template, which also removes thinking trajectory from the SFT data.

Here are the exact configurations for each training framework:

TRL:
```

model_name_or_path: /deepseek-ai/DeepSeek-R1-Distill-Qwen-32B
model_revision: main
torch_dtype: bfloat16
attn_implementation: flash_attention_2

# Data training arguments
dataset_name: /projects/open-r1/data/s1k_1_1/train.json
dataset_num_proc: 48

#SFT hyperparam
max_length: 16384
weight_decay: 0.01
optim: adamw_torch
lr_scheduler_type: cosine
warmup_ratio: 0
learning_rate: 1.0e-05
completion_only_loss: true
gradient_accumulation_steps: 2
per_device_eval_batch_size: 1
per_device_train_batch_size: 1

# SFT trainer config
max_steps: -1
num_train_epochs: 5
bf16: true
do_eval: false
use_liger_kernel: true
eval_strategy: 'no'
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
hub_model_id: DeepSeek-R1-Distill-Qwen-32B_s1k_1_1
hub_strategy: every_save
log_level: info
logging_steps: 1
logging_strategy: steps
packing: false
output_dir: data/DeepSeek-R1-Distill-Qwen-32B_sft_s1k_ds_data-orig_chat_template
overwrite_output_dir: true
push_to_hub: false
report_to:
- wandb
save_strategy: "steps"
save_steps: "100000"
seed: 42
```

LLama-factory:
```
llamafactory-cli train \
      --stage sft \
      --do_train True \
      --model_name_or_path /models/deepseek-ai/DeepSeek-R1-Distill-Qwen-32B \
      --preprocessing_num_workers 16 \
      --finetuning_type full \
      --template deepseekr1 \
      --flash_attn fa2 \
      --dataset_dir train/LLaMA-Factory/data \
      --dataset s1k_deepseek \
      --cutoff_len 16384 \
      --learning_rate 1e-05 \
      --num_train_epochs 5.0 \
      --max_samples 100000 \
      --per_device_train_batch_size 1 \
      --gradient_accumulation_steps 2 \
      --lr_scheduler_type cosine \
      --max_grad_norm 1.0 \
      --logging_steps 5 \
      --save_steps 1000 \
      --warmup_steps 0 \
      --packing False \
      --report_to none \
      --output_dir saves/DeepSeek-R1-Distill-Qwen-32B/full/s1k_ds_data-ds_template \
      --bf16 True \
      --plot_loss True \
      --trust_remote_code True \
      --ddp_timeout 180000000 \
      --include_num_input_tokens_seen True \
      --optim adamw_torch \
      --deepspeed train/LLaMA-Factory/examples/deepspeed/ds_z3_offload_config.json
  ```

Has anyone else encountered this issue?

**Comment by ASoleimaniB** [comment]
Do you have any updates on this?

**Comment by jiosephlee** [comment]
What version of trl are you using and what do the loss plots look like?

**Comment by Fazzioni** [comment]
I've had some issues with TRL as well, both with GRPO with reward collapse and experiments with distillation and DPO, the results seem to be degraded.

**Comment by albertvillanova** [comment]
Hey @Hasuer, sorry for the very slow reply — we're catching up on the backlog. 

Unfortunately there isn't enough here for us to investigate: there's no `trl env` output, no TRL version, and no loss curves on either side, so we have no way to tell whether the gap comes from the chat template (DeepSeek-R1-Distill's default vs LLaMA-Factory's `deepseekr1` template — these are not equivalent in how they handle the `<think>` block and trailing newlines), `completion_only_loss` masking, Liger kernel vs eager (you enabled it on the TRL side only), the DeepSpeed ZeRO-3 + CPU-offload config that only LLaMA-Factory uses, or something else entirely. 

I'm going to close this as not actionable, but happy to re-open if you can share a minimal repro with `trl env` output, the TRL version you ran, and ideally a quick comparison of the train loss curves between the two runs — that usually makes the divergence point obvious.

**Comment by qgallouedec** [comment]
could be related: https://github.com/huggingface/trl/issues/5138, this issue is better documented
